-1

Tested at: be1e263f2d4ad593d5061d1a66c8b3897b6ecfd4


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/3043808cc0a269062112e10ede407ed14759d8d3
https://github.com/cms-sw/cmssw/commit/72fcaf941a002f947b0e3ac0d20311b847afa4f4
https://github.com/cms-sw/cmssw/commit/c8d9643e777d6e8e12eb48e8db312497c5e76ba8
https://github.com/cms-sw/cmssw/commit/6aaa765fc6595b89512b4947b5eed7d6cc22299d
https://github.com/cms-sw/cmssw/commit/d0cc21f6f7f18d179e626f33041dde5f512e7c52
https://github.com/cms-sw/cmssw/commit/aa14a75177e9f010b3b8d0af0440598735c49311
https://github.com/cms-sw/cmssw/commit/484eb1496c7066c869edd7acf315871de36755bf
https://github.com/cms-sw/cmssw/commit/505e1cc0dd612130d07dadd4dabdbb69d2939ee6
https://github.com/cms-sw/cmssw/commit/88641d28cc1869081fbca832fdd464fe44a38d50
https://github.com/cms-sw/cmssw/commit/b9e49920ac6fb7c8b41a37df4151eae9de6776ea
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20409/22763/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20409/22763/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20409/22763/summary.html

I found follow errors while testing this PR

**Failed tests:  Build  ClangBuild**

* **Build**:

I found an error when building: 
 
 <pre>/cvmfs/cms-ib.cern.ch/nweek-02488/slc6_amd64_gcc630/external/thepeg/2.0.4-ghjeda4/include/ThePEG/Pointer/RCPtr.h: In instantiation of 'ThePEG::Pointer::RCPtr<T>& ThePEG::Pointer::RCPtr<T>::create(ThePEG::Pointer::RCPtr<T>::const_reference) [with T = ThePEG::LesHouchesInterface; ThePEG::Pointer::RCPtr<T>::const_reference = const ThePEG::LesHouchesInterface&]':
/cvmfs/cms-ib.cern.ch/nweek-02488/slc6_amd64_gcc630/external/thepeg/2.0.4-ghjeda4/include/ThePEG/Pointer/RCPtr.h:126:22:   required from 'static ThePEG::Pointer::RCPtr<T> ThePEG::Pointer::RCPtr<T>::Create(ThePEG::Pointer::RCPtr<T>::const_reference) [with T = ThePEG::LesHouchesInterface; ThePEG::Pointer::RCPtr<T>::const_reference = const ThePEG::LesHouchesInterface&]'
/cvmfs/cms-ib.cern.ch/nweek-02488/slc6_amd64_gcc630/external/thepeg/2.0.4-ghjeda4/include/ThePEG/Pointer/RCPtr.h:869:69:   required from 'static ThePEG::Pointer::PtrTraits<ThePEG::Pointer::RCPtr<T> >::pointer ThePEG::Pointer::PtrTraits<ThePEG::Pointer::RCPtr<T> >::create(ThePEG::Pointer::PtrTraits<ThePEG::Pointer::RCPtr<T> >::const_reference) [with T = ThePEG::LesHouchesInterface; ThePEG::Pointer::PtrTraits<ThePEG::Pointer::RCPtr<T> >::pointer = ThePEG::Pointer::RCPtr<ThePEG::LesHouchesInterface>; ThePEG::Pointer::PtrTraits<ThePEG::Pointer::RCPtr<T> >::const_reference = const ThePEG::LesHouchesInterface&]'
/cvmfs/cms-ib.cern.ch/nweek-02488/slc6_amd64_gcc630/external/thepeg/2.0.4-ghjeda4/include/ThePEG/Pointer/PtrTraits.h:204:55:   required from 'typename ThePEG::Pointer::Ptr<T>::pointer ThePEG::Pointer::new_ptr(const T&) [with T = ThePEG::LesHouchesInterface; typename ThePEG::Pointer::Ptr<T>::pointer = ThePEG::Pointer::RCPtr<ThePEG::LesHouchesInterface>]'
/build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_3_X_2017-09-05-2300/src/GeneratorInterface/ThePEGInterface/plugins/ThePEGLesHouchesInterface.cc:38:53:   required from here
/cvmfs/cms-ib.cern.ch/nweek-02488/slc6_amd64_gcc630/external/thepeg/2.0.4-ghjeda4/include/ThePEG/Pointer/RCPtr.h:147:11: error: invalid new-expression of abstract class type 'ThePEG::LesHouchesInterface'
     ptr = new T(t);
           ^~~~~~~~
/build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_3_X_2017-09-05-2300/src/GeneratorInterface/ThePEGInterface/plugins/ThePEGLesHouchesInterface.cc:30:7: note:   because the following virtual functions are pure within 'ThePEG::LesHouchesInterface':
 class LesHouchesInterface : public LesHouchesReader {
       ^~~~~~~~~~~~~~~~~~~
</pre>

* **Clang**:

I found a compilation error while trying to compile with clang: 
I used this command: 
 scram b vclean && scram build -k -j 16 USER_CXXFLAGS='-fsyntax-only' COMPILER='llvm compile'


 <pre>>> Compiling edm plugin /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_3_X_2017-09-05-2300/src/DQMOffline/Trigger/plugins/TopMonitor.cc 
>> Compiling edm plugin /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_3_X_2017-09-05-2300/src/DQMOffline/Trigger/plugins/MuonMonitor.cc 
In file included from /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_3_X_2017-09-05-2300/src/GeneratorInterface/ThePEGInterface/src/ThePEGInterface.cc:22:
In file included from /cvmfs/cms-ib.cern.ch/nweek-02488/slc6_amd64_gcc630/external/thepeg/2.0.4-ghjeda4/include/ThePEG/Repository/Repository.h:16:
In file included from /cvmfs/cms-ib.cern.ch/nweek-02488/slc6_amd64_gcc630/external/thepeg/2.0.4-ghjeda4/include/ThePEG/PDT/ParticleData.h:19:
/cvmfs/cms-ib.cern.ch/nweek-02488/slc6_amd64_gcc630/external/thepeg/2.0.4-ghjeda4/include/ThePEG/Utilities/Selector.h:131:50: error: ISO C++1z does not allow dynamic exception specifications [-Wdynamic-exception-spec]
  T & select(double rnd, double * remainder = 0) throw(range_error);
                                                 ^~~~~~~~~~~~~~~~~~
/cvmfs/cms-ib.cern.ch/nweek-02488/slc6_amd64_gcc630/external/thepeg/2.0.4-ghjeda4/include/ThePEG/Utilities/Selector.h:131:50: note: use 'noexcept(false)' instead
  T & select(double rnd, double * remainder = 0) throw(range_error);
                                                 ^~~~~~~~~~~~~~~~~~
</pre>


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/3043808cc0a269062112e10ede407ed14759d8d3
https://github.com/cms-sw/cmssw/commit/72fcaf941a002f947b0e3ac0d20311b847afa4f4
https://github.com/cms-sw/cmssw/commit/c8d9643e777d6e8e12eb48e8db312497c5e76ba8
https://github.com/cms-sw/cmssw/commit/6aaa765fc6595b89512b4947b5eed7d6cc22299d
https://github.com/cms-sw/cmssw/commit/d0cc21f6f7f18d179e626f33041dde5f512e7c52
https://github.com/cms-sw/cmssw/commit/aa14a75177e9f010b3b8d0af0440598735c49311
https://github.com/cms-sw/cmssw/commit/484eb1496c7066c869edd7acf315871de36755bf
https://github.com/cms-sw/cmssw/commit/505e1cc0dd612130d07dadd4dabdbb69d2939ee6
https://github.com/cms-sw/cmssw/commit/88641d28cc1869081fbca832fdd464fe44a38d50
https://github.com/cms-sw/cmssw/commit/b9e49920ac6fb7c8b41a37df4151eae9de6776ea
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20409/22763/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20409/22763/git-merge-result

--- a/FastSimulation/CaloGeometryTools/interface/Crystal.h
+++ b/FastSimulation/CaloGeometryTools/interface/Crystal.h
@@ -31,7 +31,7 @@ class Crystal
   /// Empty constructor 
   Crystal(){number_ = 0;};
   /// constructor from DetId
-  Crystal(const DetId&  cell,const BaseCrystal* bc=0);
+  Crystal(const DetId&  cell,const BaseCrystal* bc=nullptr);
 
   /// get the i-th corner
   inline const XYZPoint& getCorner(unsigned i) const { return myCrystal_->getCorner(i);};

--- a/FastSimulation/CaloGeometryTools/src/CaloGeometryHelper.cc
+++ b/FastSimulation/CaloGeometryTools/src/CaloGeometryHelper.cc
@@ -42,7 +42,7 @@ void CaloGeometryHelper::initialize(double bField)
   buildCrystalArray();
   buildNeighbourArray();
   bfield_ = bField;
-  preshowerPresent_=(getEcalPreshowerGeometry()!=0);
+  preshowerPresent_=(getEcalPreshowerGeometry()!=nullptr);
     
   if(preshowerPresent_)
     {
@@ -353,7 +353,7 @@ bool CaloGeometryHelper::simplemove(DetId& cell, const CaloDirection& dir) const
   else if(cell.subdetId()==EcalEndcap)
     neighbours= EcalEndcapTopology_->getNeighbours(cell,dir);
   
-  if(neighbours.size()>0 && !neighbours[0].null())
+  if ((!neighbours.empty()) && (!neighbours[0].null()))
     {
       cell = neighbours[0];
       return true;
@@ -460,7 +460,7 @@ void CaloGeometryHelper::buildCrystalArray()
   //std::cout << " Building the array of crystals (barrel) " ;
   const std::vector<DetId>&  vec(EcalBarrelGeometry_->getValidDetIds(DetId::Ecal,EcalBarrel));
   unsigned size=vec.size();    
-  const CaloCellGeometry * geom=0;
+  const CaloCellGeometry * geom=nullptr;
   for(unsigned ic=0; ic<size; ++ic) 
     {
       unsigned hashedindex=EBDetId(vec[ic]).hashedIndex();

--- a/FastSimulation/CaloGeometryTools/src/DistanceToCell.cc
+++ b/FastSimulation/CaloGeometryTools/src/DistanceToCell.cc
@@ -4,7 +4,7 @@
 #include "Geometry/CaloGeometry/interface/CaloCellGeometry.h"
 #include "Geometry/HcalTowerAlgo/interface/HcalGeometry.h"
 
-DistanceToCell::DistanceToCell():det_(0) {;}
+DistanceToCell::DistanceToCell():det_(nullptr) {;}
 
 DistanceToCell::DistanceToCell(const DistanceToCell& dist)
 {

